content/courses/advanced-ada/chapters/numerics.rst

```diff
@@ -1639,12 +1639,747 @@ precision, 0.2774 is rounded to 0.277, and 0.2777 is rounded to 0.278.
+       BI : Big_Integer;
+    begin
+       BI := 12345678901234567890;
+
+       Put_Line ("BI: " & BI'Image);
+
+       BI := -@ + BI / 2;
+       BI :=  @ - BI * 2;
+
+       Put_Line ("BI: " & BI'Image);
+    end Show_Simple_Big_Numbers_Operators;
+
+In this example, we're applying the four basic operators (:ada:`+`, :ada:`-`,
+:ada:`*`, :ada:`/`) on big integers.
+
+
+Factorial
+~~~~~~~~~
+
+A typical example is the :wikipedia:`factorial <Factorial>`: a sequence of the
+factorial of consecutive small numbers can quickly lead to big numbers. Let's
+take this implementation as an example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Numerics.Factorial_Integer
+    :class: ada-run-expect-failure
+
+    function Factorial (N : Integer)
+                        return Long_Long_Integer;
+
+    function Factorial (N : Integer)
+                        return Long_Long_Integer is
+       Fact : Long_Long_Integer := 1;
+    begin
+       for I in 2 .. N loop
+          Fact := Fact * Long_Long_Integer (I);
+       end loop;
+
+       return Fact;
+    end Factorial;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with Factorial;
+
+    procedure Show_Factorial is
+    begin
+       for I in 1 .. 50 loop
+          Put_Line (I'Image & "! = "
+                    & Factorial (I)'Image);
+       end loop;
+    end Show_Factorial;
+
+Here, we're using :ada:`Long_Long_Integer` for the computation and return type
+of the :ada:`Factorial` function. The last number we're able to calculate
+before getting an exception is `20!`, which basically shows the limitation of
+standard integers for this kind of algorithms. If we use big integers instead,
+we can easily display all numbers up to `50!` (and more!):
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Numerics.Factorial_Big_Numbers
+
+    pragma Ada_2022;
+
+    with Ada.Numerics.Big_Numbers.Big_Integers;
+    use  Ada.Numerics.Big_Numbers.Big_Integers;
+
+    function Factorial (N : Integer)
+                        return Big_Integer;
+
+    function Factorial (N : Integer)
+                        return Big_Integer is
+       Fact : Big_Integer := 1;
+    begin
+       for I in 2 .. N loop
+          Fact := Fact * To_Big_Integer (I);
+       end loop;
+
+       return Fact;
+    end Factorial;
+
+    pragma Ada_2022;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with Factorial;
+
+    procedure Show_Big_Number_Factorial is
+    begin
+       for I in 1 .. 50 loop
+          Put_Line (I'Image & "! = "
+                    & Factorial (I)'Image);
+       end loop;
+    end Show_Big_Number_Factorial;
+
+As we can see in this example, replacing the :ada:`Long_Long_Integer` type by
+the :ada:`Big_Integer` type fixes the problem (the runtime exception) that we
+had in the previous version.
+(Note that we're using the :ada:`To_Big_Integer` function to convert from
+:ada:`Integer` to :ada:`Big_Integer`: we discuss these conversions next.)
+
```

Comment:
Are you going to mention that there is a limit to the upper bounds for big integers, but that it is not dependent on the hardware types. In other words the compiler can decide how much memory it wants to use to represent them...

Reply:
Good point, I've just paraphrased what you said here.